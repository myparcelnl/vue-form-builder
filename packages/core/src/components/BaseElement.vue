<template>
  <component
    :is="element.component"
    v-bind="element.props"
    v-on="hooks" />
</template>

<script lang="ts">
import {PropType, computed, defineComponent, toRefs} from 'vue';
import {PlainElementInstance} from '../form';

export default defineComponent({
  name: 'BaseElement',

  props: {
    element: {
      type: Object as PropType<PlainElementInstance>,
      required: true,
    },
  },

  setup: (props) => {
    const propRefs = toRefs(props);

    return {
      propRefs,
      hooks: computed(() => {
        return props.element.hooks.registeredHooks.reduce(
          (acc, hook) => ({
            ...acc,
            [hook.name]: hook.callback,
          }),
          {},
        );
      }),
    };
  },
});
</script>
