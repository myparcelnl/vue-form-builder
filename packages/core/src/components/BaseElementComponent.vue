<template>
  <component
    :is="element.component"
    v-bind="element.props"
    v-on="hooks" />
</template>

<script lang="ts">
import {PropType, computed, defineComponent, toRefs} from 'vue';
import {BaseField} from '../services';

export default defineComponent({
  name: 'BaseElementComponent',

  props: {
    element: {
      type: Object as PropType<BaseField>,
      required: true,
    },
  },

  setup: (props) => {
    const propRefs = toRefs(props);

    return {
      propRefs,
      hooks: computed(() => {
        return props.element.hooks.registeredHooks.reduce(
          (acc, hook) => ({
            ...acc,
            [hook.name]: hook.callback,
          }),
          {},
        );
      }),
    };
  },
});
</script>
