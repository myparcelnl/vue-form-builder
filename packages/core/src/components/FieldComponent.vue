<template>
  <span
    v-show="field.isVisible"
    :class="field.isVisible ? field.form.config.fieldClass : null">
    <component
      :is="field.component"
      v-if="field.lazy"
      v-model.lazy="/* eslint-disable vue/no-mutating-props */ field.ref"
      v-bind="bindData"
      v-on="hooks" />

    <component
      :is="field.component"
      v-else
      v-model="/* eslint-disable vue/no-mutating-props */ field.ref"
      v-bind="bindData"
      v-on="hooks" />

    <!--    <span class="pointer-events-none relative z-10"> -->
    <!--      <span -->
    <!--        class="absolute bg-pink-200 dark:bg-pink-800 duration-100 group-hover:opacity-100 opacity-0 p-2 right-0 rounded-lg text-xs transition-opacity whitespace-nowrap"> -->
    <!--        name: {{ field.name }} -->
    <!--        <br /> -->
    <!--        value: {{ field.ref }} -->

    <!--        <CodeBlock -->
    <!--          :code="{ -->
    <!--            value: field.ref, -->
    <!--            isDirty: field.isDirty, -->
    <!--            isDisabled: field.isDisabled, -->
    <!--            isOptional: field.isOptional, -->
    <!--            isSuspended: field.isSuspended, -->
    <!--            isTouched: field.isTouched, -->
    <!--            isValid: field.isValid, -->
    <!--            isVisible: field.isVisible, -->
    <!--          }" -->
    <!--          hide-controls /> -->
    <!--      </span> -->
    <!--    </span> -->
  </span>
</template>

<script lang="ts">
import {PropType, computed, defineComponent, toRefs} from 'vue';
import {CodeBlock} from '@myparcel/vue-form-builder-shared';
import {FieldInstance} from '../services';
import {useLifeCycleHooks} from '../composables/useLifeCycleHooks';

export default defineComponent({
  name: 'FieldComponent',
  components: {
    CodeBlock,
  },

  props: {
    field: {
      type: Object as PropType<FieldInstance>,
      required: true,
    },
  },

  setup: (props) => {
    const propRefs = toRefs(props);

    useLifeCycleHooks(props.field.hooks, props.field);

    const hooks = computed(() => {
      return props.field.hooks.registeredHooks.reduce(
        (acc, hook) => {
          if (!hook.name.startsWith('on')) {
            return acc;
          }

          return {
            ...acc,
            [hook.name.replace(/^on/, '')]: (value) => hook.callback(props.field, value),
          };
        },
        {
          blur: props.field.blur,
          focus: props.field.focus,
          click: props.field.click,
        },
      );
    });

    // if (props.field.hooks.has('mounted')) {
    //   onMounted(async () => {
    //     await props.field.hooks.execute('mounted', props.field);
    //   });
    // }

    // onBeforeUpdate(async () => {
    //   await props.field.hooks.execute('beforeUpdate', props.field);
    // });
    //
    // onUpdated(async () => {
    //   await props.field.hooks.execute('updated', props.field);
    // });
    //
    // onActivated(async () => {
    //   await props.field.hooks.execute('activated', props.field);
    // });
    //
    // onDeactivated(async () => {
    //   await props.field.hooks.execute('deactivated', props.field);
    // });
    //
    // onBeforeUnmount(async () => {
    //   await props.field.hooks.execute('beforeUnmount', props.field);
    // });
    //
    // onUnmounted(async () => {
    //   await props.field.hooks.execute('unmounted', props.field);
    // });

    return {
      propRefs,
      hooks,
      bindData: computed(() => {
        return {
          ...(props.field.props ?? {}),
          id: props.field.id ?? props.field.name,
          label: props.field.label,
          name: props.field.name,
          disabled: props.field.isDisabled,
          valid: props.field.isValid,
        };
      }),
    };
  },
});
</script>
