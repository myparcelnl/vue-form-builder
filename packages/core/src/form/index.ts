export * from './Form';
export * from './Form.types';
export * from './defineField';
export * from './defineForm';
export * from './interactive-element';
export * from './plain-element';
