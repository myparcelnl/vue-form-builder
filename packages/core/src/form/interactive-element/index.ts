export * from './InteractiveElement';
export * from './InteractiveElement.types';
export * from './defineField';
export * from './validator';
