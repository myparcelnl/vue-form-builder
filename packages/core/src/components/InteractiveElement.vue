<template>
  <div
    v-show="element.isVisible"
    :class="element.isVisible ? element.form.config.fieldClass : null">
    <component
      :is="element.component"
      v-if="element.lazy"
      v-model.lazy="/* eslint-disable vue/no-mutating-props */ element.ref"
      v-bind="bindData"
      v-on="hooks" />

    <component
      :is="element.component"
      v-else
      v-model="/* eslint-disable vue/no-mutating-props */ element.ref"
      v-bind="bindData"
      v-on="hooks" />
  </div>
</template>

<script lang="ts">
import {PropType, computed, defineComponent, provide, toRefs} from 'vue';
import {INJECT_ELEMENT} from '../services';
import {InteractiveElementInstance} from '../../../form-builder/src';
import {useLifeCycleHooks} from '../composables';

export default defineComponent({
  name: 'InteractiveElement',

  props: {
    element: {
      type: Object as PropType<InteractiveElementInstance>,
      required: true,
    },
  },

  setup: (props) => {
    const lifecycleHooks = useLifeCycleHooks();
    provide(INJECT_ELEMENT, props.element);

    const propRefs = toRefs(props);

    lifecycleHooks.register(propRefs.element.value.hooks, propRefs.element);

    const hooks = computed(() => {
      return props.element.hooks.registeredHooks.reduce(
        (acc, hook) => {
          if (!hook.name.startsWith('on')) {
            return acc;
          }

          return {
            ...acc,
            // todo: fix types
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            [hook.name.replace(/^on/, '')]: (value: any) => hook.callback(props.element, value),
          };
        },
        // TODO: fix types
        {
          blur: props.element.blur,
          focus: props.element.focus,
          click: props.element.click,
        },
      );
    });

    return {
      propRefs,
      hooks,
      bindData: computed(() => {
        return {
          ...(props.element.props ?? {}),
          id: props.element.name ?? props.element.name,
          label: props.element.label,
          name: props.element.name,
          disabled: props.element.isDisabled,
          valid: props.element.isValid,
          warnings: props.element.errors,
        };
      }),
    };
  },
});
</script>
