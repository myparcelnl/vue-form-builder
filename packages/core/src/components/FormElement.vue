<template>
  <template v-if="field.teleportSelector">
    <Teleport
      v-if="isLoaded"
      :to="field.teleportSelector">
      <InteractiveElement
        v-show="field.isVisible"
        v-if="field.hasOwnProperty('ref')"
        :element="field" />
      <PlainElement
        v-else
        v-show="field.isVisible"
        :element="field" />
    </Teleport>
  </template>
  <template v-else>
    <InteractiveElement
      v-show="field.isVisible"
      v-if="field.hasOwnProperty('ref')"
      :element="field" />
    <PlainElement
      v-else
      v-show="field.isVisible"
      :element="field" />
  </template>
</template>

<script lang="ts">
import {PropType, defineComponent, onMounted, ref} from 'vue';
import {AnyElementInstance} from '../types';
import InteractiveElement from './InteractiveElement.vue';
import PlainElement from './PlainElement.vue';

export default defineComponent({
  name: 'FormElement',
  components: {InteractiveElement, PlainElement},
  props: {
    field: {
      type: Object as PropType<AnyElementInstance>,
      required: true,
    },
  },

  setup: () => {
    const isLoaded = ref(false);

    onMounted(() => {
      isLoaded.value = true;
    });

    return {
      isLoaded,
    };
  },
});
</script>
