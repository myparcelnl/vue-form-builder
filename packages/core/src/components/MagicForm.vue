<template>
  <form
    :id="form.name"
    :class="[
      form.config.formClass,
      {
        valid: form.isValid,
        invalid: !form.isValid,
      },
    ]"
    @submit.prevent="form.submit">
    <template
      v-for="(field, index) in fields"
      :key="`field--${field.name ?? 'unnamed'}--${index}`">
      <InteractiveElement
        v-show="field.isVisible"
        v-if="field.hasOwnProperty('ref')"
        :element="field" />

      <PlainElement
        v-else
        v-show="field.isVisible"
        :element="field" />
    </template>
  </form>
</template>

<script lang="ts">
import {PropType, defineComponent, provide, toRef} from 'vue';
import {FormInstance} from '../form';
import {INJECT_FORM} from '../services';
import InteractiveElement from './InteractiveElement.vue';
import PlainElement from './BaseElement.vue';
import {useLifeCycleHooks} from '../composables';

export default defineComponent({
  name: 'MagicForm',
  components: {
    InteractiveElement,
    PlainElement,
  },

  props: {
    form: {
      type: Object as PropType<FormInstance>,
      required: true,
    },
  },

  setup: (props) => {
    provide(INJECT_FORM, props.form);

    const lifeCycleHooks = useLifeCycleHooks();

    lifeCycleHooks.register(props.form.hooks, props.form);

    return {
      fields: toRef(props.form, 'fields'),
    };
  },
});
</script>
