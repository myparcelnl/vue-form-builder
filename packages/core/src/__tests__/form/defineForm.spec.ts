import {defineForm} from '../../form';
import {ref} from 'vue';
import {afterAll, beforeAll, describe, expect, it, SpyInstance} from 'vitest';
import { DOMWrapper, flushPromises, mount } from '@vue/test-utils';
import TextInput from '../elements/TextInput.vue';
import { vi } from 'vitest';
import { FormInstance } from '../../../lib';
import { InteractiveElement } from '../../form/interactive-element';
import { MagicForm } from '../../components';

const formIsInvalid = (formElement: DOMWrapper<HTMLElement>, form: FormInstance) => {
  expect(form.isValid.value).toBe(false);
  expect(formElement.classes()).toContain('invalid');
  expect(formElement.classes()).not.toContain('valid');
};

const formIsValid = (formElement: DOMWrapper<HTMLElement>, form: FormInstance) => {
  expect(form.isValid.value).toBe(true);
  expect(formElement.classes()).toContain('valid');
  expect(formElement.classes()).not.toContain('invalid');
}

describe('Form Generation', () => {
  let consoleSpy: SpyInstance;
  beforeAll(() => {
    consoleSpy = vi.spyOn(console, 'info').mockImplementation(() => {});
  });

  afterAll(() => {
    consoleSpy.mockRestore();
  });
  describe.skip('form builder', () => {
    const form = defineForm('test', {
      fields: [
        {
          name: 'named',
          component: 'input',
        },
        {
          component: 'br',
        },
        {
          name: 'text',
          component: TextInput,
          ref: ref('initial'),
        },
      ],
    });
    it('works', () => {
      // expect(form.model.named).toBeInstanceOf(InteractiveElement);
      expect(form.model.text).toBeInstanceOf(InteractiveElement);

      expect(form.model.text.ref).toEqual('initial');

      form.model.text.ref = 'changed';
      expect(form.model.text.ref).toEqual('changed');

      // expect(form.registeredHooks).toEqual([]);
      // expect(form.fields).toEqual({
      //   toggle: {
      //     component: CustomCheckbox,
      //   },
      // });
    });

    it('can use vue element wrapper', async () => {
      const wrapper = mount(MagicForm, {
        props: {
          form,
        },
      });
      await flushPromises();
      const formElement = wrapper.find('form');
      expect(formElement.exists()).toBe(true);
      expect(formElement.attributes('id')).toBe('test');
      expect(formElement.find('input').exists()).toBe(true);
      expect(formElement.find('br').exists()).toBe(true);
      expect(formElement.find('input[name="text"]').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('basic validation', () => {
    it('can determine if a text input is valid based on single predicate', async () => {
      const firstName = ref('');
      const lastName = ref('');
      const validationForm = defineForm('validationForm', {
        fields: [
          {
            name: 'firstName',
            component: TextInput,
            ref: firstName,
            validate: (_, value) => String(value).startsWith('J'),
            errorMessage: 'Field must start with "J"',
          },
          {
            name: 'lastName',
            component: TextInput,
            ref: lastName,
          },
        ],
      });
      const wrapper = mount(MagicForm, {
        props: {
          form: validationForm,
        },
      });
      await flushPromises();
      const formElement = wrapper.find('form');

      // expect default state to be valid regardless of input
      formIsValid(formElement, validationForm);

      const firstNameInput = wrapper.find('input[name="firstName"]');
      await firstNameInput.setValue('Karen');
      expect(firstName.value).toBe('Karen');

      const lastNameInput = wrapper.find('input[name="lastName"]');
      await lastNameInput.setValue('Doe');
      expect(lastName.value).toBe('Doe');

      await formElement.trigger('submit');
      await flushPromises();
      formIsInvalid(formElement, validationForm);
    });
  });
})
