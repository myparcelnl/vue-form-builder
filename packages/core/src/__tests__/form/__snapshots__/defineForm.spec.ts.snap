// Vitest Snapshot v1

exports[`Form Generation > form builder > can use vue element wrapper 1`] = `
"<form id=\\"test\\" class=\\"valid\\"><input><br>
  <div class=\\"\\"><input id=\\"text\\" name=\\"text\\" warnings=\\"\\" valid=\\"[object Promise]\\"></div>
</form>"
`;
